[master < T0663-MG] Add kafka documentation (#163)
memgraph/docs: docs/transformation-modules/api/python-api.md

---
id: python-api
title: Transformations Python API
sidebar_label: Python API
---

This complementary API documentation for `mgp.py`  which contains definitions of the public Transformation Python API provided by Memgraph.
At the core, this is a wrapper around the **[C API](c-api.md)**. This source file can be found in the Memgraph installation directory,
under `python_support`. On the standard Debian installation, this will be under `/usr/lib/memgraph/python_support`.

:::caution
**NOTE:** This part of the documentation is still under development. An updated version will soon be available.
:::

## `mgp.transformation(func)`
Register func as a transformation of the current module.

> transformation is meant to be used as a decorator function to register a transformation 
> within the current module. The registered func needs to be a callable which either accepts
> a single argument of type Messages or two arguments of type TransCtx and Messages respectively. 
> The return type is fixed and it should be of the form:
> Record(query=str, parameters=mgp.Map). Multiple records can be
> produced by returning an iterable of them. Registering generator functions
> is currently not supported.

Reviewer: ```suggestion
`mgp.transformation` is meant to be used as a decorator function to register a transformation 
within the current module. The registered `func` needs to be a callable which either accepts
a single argument of type `Messages` or two arguments of type `TransCtx` and `Messages` respectively. 
The return type is fixed and it should be of the form:
`Record(query=str, parameters=mgp.Map)`. Multiple records can be
produced by returning an iterable of them. Registering generator functions
is currently not supported.
```

And do the same every time with the backtick when you see something which is code specific in the middle of the text.

### Examples

    return result_queries
```python
    import mgp

    @mgp.transformation
    def transformation(context: mgp.TransCtx,
                       messages: mgp.Messages
                       ) -> mgp.Record(query=str, parameters=mgp.Map):
    result_queries = []

    for i in range(0, messages.total_messages()):
        message = messages.message_at(i)
        payload_as_str = message.payload().decode("utf-8")
        result_queries.append(mgp.Record(
            query=f"CREATE (n:MESSAGE {{timestamp: '{message.timestamp()}', payload: '{payload_as_str}', topic: '{message.topic_name()}'}})",
            parameters={"param1": "value1", "param2": 2}))
  
    return result_queries
```
This transformation extracts the interesting members of each message and stores them in a Record of query which wraps a CREATE clause 
with all the interesting members (timestamp, payload, etc) and parameters, its parameters.

Any errors can be reported by raising an Exception.

## `class mgp.Message(messages)`
Bases: `object`

Represents a single message. 
You should not globally store Message.

### `is_valid()`

> Returns true if self is valid and can be accessed.

Reviewer: ```suggestion
Returns true if message is valid and can be accessed.
```

### `payload()`
Returns the payload of the message.
Raises an InvalidMessageError if is_valid() is false.

### `topic_name()`
Returns the topic_name of the message.
Raises an InvalidMessageError if is_valid() is false.

### `key()`
Returns the key of the message as bytes
Raises an InvalidMessageError if is_valid() is false.

### `timestamp()`
Returns the timestamp of the message.
Raises an InvalidMessageError if is_valid() is false.

## `class mgp.Messages(messages)`
Bases: `object`

Represents a list of messages passed to a transformation. 
You should not globally store messages.

### `is_valid()`

> Returns true if self is valid and can be accessed.

Reviewer: ```suggestion
Returns true if messages object is valid and can be accessed.
```

### `total_messages()`
Returns the number of messages contained.
Raises InvalidMessagesError if is_valid() is false.

### `message_at(id)`

> Returns the underline message at index id.

Reviewer: ```suggestion
Returns the underlying message at index `id`.
```
